src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Refactoring/ExtractToCodeBehindCodeActionResolver.cs

```diff
@@ -0,0 +1,156 @@
+using Microsoft.AspNetCore.Razor.Language.Intermediate;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.AspNetCore.Razor.LanguageServer.ProjectSystem;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Razor;
+using Newtonsoft.Json.Linq;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+
+using CSharpSyntaxFactory = Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Refactoring
+{
+    class ExtractToCodeBehindCodeActionResolver : RazorCodeActionResolver
+    {
+        private readonly ForegroundDispatcher _foregroundDispatcher;
+        private readonly DocumentResolver _documentResolver;
+
+        public override string Action => "ExtractToCodeBehind";
+
+        public ExtractToCodeBehindCodeActionResolver(
+            ForegroundDispatcher foregroundDispatcher,
+            DocumentResolver documentResolver)
+        {
+            if (foregroundDispatcher is null)
+            {
+                throw new ArgumentNullException(nameof(foregroundDispatcher));
+            }
+
+            if (documentResolver is null)
+            {
+                throw new ArgumentNullException(nameof(documentResolver));
+            }
+
+            _foregroundDispatcher = foregroundDispatcher;
+            _documentResolver = documentResolver;
+        }
+
+        override public async Task<WorkspaceEdit> ResolveAsync(JObject data, CancellationToken cancellationToken)
+        {
+            var actionParams = data.ToObject<ExtractToCodeBehindParams>();
+            var path = actionParams.Uri.LocalPath;
+
+            var document = await Task.Factory.StartNew(() =>
+            {
+                _documentResolver.TryResolveDocument(path, out var documentSnapshot);
+                return documentSnapshot;
+            }, cancellationToken, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler);
+            if (document is null)
+            {
+                return null;
+            }
+
+            var text = await document.GetTextAsync();
+            if (text is null)
+            {
+                return null;
+            }
+
+            var codeDocument = await document.GetGeneratedOutputAsync();
+            if (codeDocument.IsUnsupported())
+            {
+                return null;
+            }
+
+            var codeBehindPath = "";
+            var n = 0;
+            do
+            {
+                var identifier = n > 0 ? n.ToString() : "";  // Make it look nice
+                codeBehindPath = Path.Combine(
+                    Path.GetDirectoryName(path),
+                    $"{Path.GetFileNameWithoutExtension(path)}{identifier}{Path.GetExtension(path)}.cs");
+                n++;
+            } while (File.Exists(codeBehindPath));
+
+            var codeBehindUri = new Uri(codeBehindPath);
```

Comment:
Use UriBuilder to avoid it inferring the incorrect Uri: https://github.com/dotnet/aspnetcore-tooling/blob/12778807aafb6b2505cec119f6562726e8431cf7/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorDiagnosticsPublisher.cs#L239-L244